packages/openlmis/src/Utils.js

```diff
@@ -0,0 +1,98 @@
+import { composeNextState } from '@openfn/language-common';
+import {
+  request as commonRequest,
+  logResponse,
+  makeBasicAuthHeader,
+} from '@openfn/language-common/util';
+
+export const authorize = state => {
+  const {
+    baseUrl,
+    access_token,
+    username,
+    password,
+    clientId = 'user-client',
+    clientSecrete = 'changeme',
+  } = state.configuration;
+
+  const headers = makeBasicAuthHeader(clientId, clientSecrete);
+
+  if (access_token) return state;
+
+  if (username && password) {
+    if (clientId === 'user-client')
+      console.warn('Using default client id:', clientId);
+    if (clientSecrete === 'changeme')
+      console.warn('Using default client secrete:', clientSecrete);
+
+    const options = {
+      query: { grant_type: 'password', username, password },
+      headers: {
+        'Content-Type': 'application/json',
+        ...headers,
+      },
+      baseUrl,
+      parseAs: 'json',
+      maxRedirections: 1,
+    };
+    return commonRequest('POST', '/api/oauth/token', options)
+      .then(response => {
+        return {
+          ...state,
+          configuration: {
+            ...state.configuration,
+            access_token: response.body.access_token,
+          },
+        };
+      })
+      .catch(err => {
+        console.error('Error authenticating with OpenLMIS');
+        console.log(err.body);
+        throw err;
+      });
+  }
+
+  // warn if no auth found
+  console.warn(
+    `WARNING: no authentication properties found on congfiguration schema!`
+  );
+  console.warn(
+    'Make sure to either set an access_token or username & password'
+  );
+};
+
+export const prepareNextState = (state, response, callback = s => s) => {
+  const { body, ...responseWithoutBody } = response;
+
+  if (!state.references) {
+    state.references = [];
+  }
+
+  const nextState = {
+    ...composeNextState(state, response.body),
+    response: responseWithoutBody,
+  };
+
+  return callback(nextState);
+};
+
+export const request = (configuration = {}, method, path, options) => {
+  const { baseUrl, access_token } = configuration;
+  const { headers = {}, ...otherOptions } = options;
+
+  const opts = {
+    parseAs: 'json',
+    baseUrl: `${baseUrl}/api`,
```

Comment:
As with FHIR, `baseUrl` should just be passed through as a baseURL (I'm SURE that undici will throw for this, is this really working? If so, why?).

You should append `/api` to the path. Although if we do this, I think you need to make it clear in the docs. Because otherwise the first thing I'll do is `get('/api/processingPeriods')` and I'll get an error.

Reply:
It's working, if i append the `/api` in `baseUrl`. If i prepend the `/api` in path. Eg `/api/${path}` It fails for operations like `get('/processingPeriods')` because the url is invalid.

Reply:
But we can fix that! See the last comment in #578
```
import nodepath from 'node:path'
...
const safepath = nodepath.join('api', path)
```
`path.join` will build a safe path with slashes in the right place

As it stands at the moment, it'll still fail if there's a trailing slash on the baseUrl in config, which isn't robust enough

Reply:
Ooh i see, okay i have adapted that approach of creating safe path using nodepath